### PR TITLE
Move domain-specific namespaces to be objects in a domain context

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
This enables a Web app to connect to more than one domain server at once.

Summary of SDK changes:
- `ContextManager` namespace item added.
- `AddressManager` and `NodesList` changed to be classes instead of namespace items.

Builds on top of PR https://github.com/vircadia/vircadia-web-sdk/pull/23.
Works in conjunction with Vircadia PR https://github.com/vircadia/vircadia/pull/1300.

Merge sequence:
- [x] PR https://github.com/vircadia/vircadia-web-sdk/pull/23.
- [ ] CR, QA.
- [ ] Merge.

#### QA

With the domain server running on localhost:
- Enable anonymous connections and restart.

In the directory where the PR files are checked out:
` `npm install` - To set up the the dev & test environment.
- `npm run lint` - Should complete with no errors.
- `npm run build` - Should complete with no errors.
- `npm run test` - All  tests should pass. 
    Run the tests against the domain server in https://github.com/vircadia/vircadia/pull/1300.
    Check that the domain-server.exe process doesn't crash and restart during the tests.
    Known issue: "Jest did not exit one second after the test run has completed." See: https://github.com/vircadia/vircadia-web-sdk/issues/21
- `npm run devdoc` : Docs should be created in `./docs/dev/` directory.
- `npm run sdkdoc` : Docs should be created in `./docs/sdk/` directory.
- Set up and run the example Web app per the README.md in the `./example` directory.
  Note: If anonymous connection isn't permitted, the app will show a "REFUSED" state.
  Known issue: State reported by `DomainServer` isn't "ERROR" if location used is invalid.

### CODE REVIEW

File: src/domain/shared/ContextManager.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+// A context, mapping class names to class instances.
+type Context = Map<string, DependencyObject>;
+
+// All the contexts.
+type ContextTable = Array<Context>;
+
+
+/*@devdoc
+ *  The <code>ContextManager</code> namespace manages sets of unique object instances. Such a set forms a "context" in which
```
Comment (digisomni) [inline src/domain/shared/ContextManager.ts:29]: What is an object instance? An object being... a DomainServer? I think the context manager and contexts in general might need more explaining as to what they are.

File: tests/domain/shared/ContextManager.unit.test.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        ContextManager.set(contextID, A);
+
+        const objectA = ContextManager.get(contextID, A);
+        expect(objectA instanceof A).toBe(true);
+
+        const objectX = ContextManager.get(contextID, A);
+        expect(objectX).toBe(objectA);
+    });
+
+    test("Can include parameters when add a dependency", () => {
```
Comment (digisomni) [inline tests/domain/shared/ContextManager.unit.test.js:44]:
```suggestion
    test("Can include parameters when adding a dependency", () => {
```

File: tests/domain/shared/ContextManager.unit.test.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        }).toThrow();
+        expect(() => {
+            ContextManager.set(contextID + 1, C);  // eslint-disable-line @typescript-eslint/restrict-plus-operands
+        }).toThrow();
+        expect(() => {
+            ContextManager.get(contextID + 1, C);  // eslint-disable-line @typescript-eslint/restrict-plus-operands
+        }).toThrow();
+    });
+
+    test("An error is thrown when try to get an object not present in the context", () => {
```
Comment (digisomni) [inline tests/domain/shared/ContextManager.unit.test.js:66]:
```suggestion
    test("An error is thrown when trying to get an object not present in the context", () => {
```

File: tests/domain/shared/ContextManager.unit.test.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    test("Can include parameters when add a dependency", () => {
+        const contextID = ContextManager.createContext();
+        ContextManager.set(contextID, B, contextID, "hello");
+        const objectB = ContextManager.get(contextID, B);
+        expect(objectB.cid).toBe(contextID);
+        expect(objectB.sov).toBe("hello");
+    });
+
+    test("Errors are thrown when try to use an invalid context ID", () => {
```
Comment (digisomni) [inline tests/domain/shared/ContextManager.unit.test.js:52]:
```suggestion
    test("Errors are thrown when trying to use an invalid context ID", () => {
```

**Review by digisomni — CHANGES_REQUESTED** [review]

**Comment by digisomni** [comment]
Two issues found during testing with PR1300.

image

image

File: src/domain/shared/ContextManager.ts (hunk 1 above)
Comment (ctrlaltdavid) [reply src/domain/shared/ContextManager.ts]: I've updated the documentation.

**Comment by ctrlaltdavid** [comment]
In those test results, the second screen shot is probably a side-effect of the problem encountered tin the first.
I cannot reproduce the problem in the first, though. Hmmm.

**Comment by digisomni** [comment]
I don't know either, it works now. Musta been a prank.

**Review by digisomni — APPROVED** [review]

**Review by Misterblue — APPROVED** [review]

**Pull request merged**